Redefine workflow result schema

mareklibra opened the pull request:
FLPATH-1678

The goal is to have unified workflow result structure suitable for both ongoing automated processing same as providing status information for the user.

mareklibra commented on shared/schemas/workflow-option-output-schema.json:
```diff
@@ -0,0 +1,18 @@
+{
```
Taken from https://github.com/parodos-dev/workflow-software-templates/blob/main/scaffolder-templates/complex-assessment-workflow/skeleton/src/main/resources/schemas/workflow-option-output-schema.json, no change

mareklibra commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,48 @@
+            "description": "The current status of a workflow execution",
+            "$ref": "workflow-status-schema.json"
+        },
+        "currentVersion": {
```
Do we still need following fields?
I would keep them for backward compatibility to simplify existing workflows' adaptation to the new schema.

pkliczewski replied:
in my opinion it is ok to remove it

mareklibra commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,48 @@
+    "properties": {
+        "status": {
+            "description": "The current status of a workflow execution",
+            "$ref": "workflow-status-schema.json"
```
Do we need workflow-specific versions as it used to be for the fields bellow?
I think we can make this structure generic-enough and mandatory.

pkliczewski replied:
workflow versioning was taken as is from Parodos. I think they do not have value in current implementation. I am ok to remove it.

mareklibra replied:
If we remove it, we can move the status "one level up" - the structure of `workflow-status-schema.json` will be content of this object

mareklibra replied:
moved one level up

mareklibra commented on shared/schemas/workflow-status-schema.json:
```diff
@@ -0,0 +1,83 @@
+            "description": "The current status of a workflow execution",
+            "type": "string",
+            "enum": [
+                "running",
```
Any other statuses?

The `running` is for not-yet finished workflows. 
What about introducing progress-info here? To get sort of info `Finished 3 of 6 tasks` but in a structured way.

pkliczewski replied:
there are cases where it is not easy to say how many tasks are done.

batzionb replied:
these are the statuses coming from sonata:
ACTIVE, ERROR, COMPLETED, ABORTED, SUSPENDED, PENDING
The status field is populated by sonata, that's why I'm not sure it's a good idea to duplicate it the new status field

mareklibra commented on shared/schemas/workflow-status-schema.json:
```diff
@@ -0,0 +1,83 @@
+    "description": "Status of workflow execution",
+    "type": "object",
+    "properties": {
+        "status": {
```
What about properties like `startedAt`, `finishedAt`?

pkliczewski replied:
good idea

masayag replied:
those values are being fetched from the data-index that stores the information about the running instances.
the workflow itself shouldn't report that information.
Generally speaking, workflow result in this context is about the application output that the user should know about, in respect to the input parameters.

Meaning, if the user provided input `X`, the workflow result should say (in case of a detected problem with `X`) if the result is a failure. This will provide an extra information, e.g error message, or an output related to the workflow context (e.g. what is the name and link of the created resources)

However, the `startedAt` and `finishedAt` will not be provided from the workflow, since it isn't the responsibility of the workflow developer to assign those values to the output structure.

In addition, there is a `status` field for the workflow run, in which case it determines how the workflow execution was completed, and based on those values, the UI offers actions to the users, such as abort, rerun or retrigger.

However, `Completed` status will be reported even for a workflow that ends with a failure and for that purpose we'd like to include the `status` in this context (at least until we'll get a new feature from sonataflow implemented with adding an indication for an error. cc @gabriel-farache - what this feature released already?)

masayag replied:
The workflow's developer can enhance https://github.com/apache/incubator-kie-kogito-runtimes/issues/3475 and add to the end state an `errorMessage` property to the metadata. By setting a value for that `errorMessage` the workflow will end with `Error` state.

However, I wouldn't like the workflow developer to report the actual expected status of a workflow, and rely on the values from the DataIndex's processInstances (aka workflow runs) for that.

See example: https://github.com/apache/incubator-kie-kogito-examples/pull/1912/files#diff-f73c639cb03eff5222587f103f21fcf8d886291ccb12c6c23ed69cd72a628188R13

gabriel-farache replied:
See https://github.com/parodos-dev/serverless-workflows/pull/135 to see how the workflow developer can set the `errorMessage` value. Not sure if setting this value also set the status to error, I'll check

gabriel-farache replied:
I confirm that is the `errorMessage` is set, the status of the workflow is set to `error` as well

mareklibra replied:
Thanks. That means we do not need the `completedWith` field. If the workflow ends with error, the `errorMessage` will be used to provide human-friendly description.

In any case, the UI will combine both the instance's error message and the result's message.

batzionb replied:
@mareklibra @gabriel-farache showed us yesterday that although the status is set to error when setting the errorMessage, the metadata.errorMessage was not propogated to the instance the UI gets
So I think we still need to confirm this

gabriel-farache replied:
I think @ydayagi has to confirm that (https://issues.redhat.com/browse/FLPATH-1674)

mareklibra replied:
Let's wait for confirmation then, I got confused with the comments in this thread.

If I read correctly, the errorMessage is meant to do that. So if not, it's a bug and we should evaluate the complexity of fixing it versus using `completedWith`.

mareklibra commented on shared/schemas/workflow-status-schema.json:
```diff
@@ -0,0 +1,83 @@
+                        "description": "Free form value of the option.",
+                        "type": "string"
+                    },
+                    "link": {
```
Would it make sense to make JSON/nested object as explicit link the link here? Is there a use-case for it?

pkliczewski replied:
as far as I remember it was added to provide references to third party systems

masayag replied:
@mareklibra I'm not sure I understand the use-case.
Do you mean to have a value with a content of a JSON?

Links are pretty common in this context (for Jira, MTA report,...). In MTA v7, the link is for downloading a zip file.
This structure doesn't specify the content of the value, it can be okay, but I'd like to verify when aligning data in the `workflow results` panel, isn't there  a need to figure out if a field is a text, number, date... for better presenting it?
I'm happy with the simplicity of this, but would like to verify it is fine from the frontend side.

mareklibra replied:
The point is to provide a way to share complex result. The UI would not parse it, just show the json, ready to copy&paste. It can be a snippet of a config file or other data object. Since this can be simulated by a `text` field, let's skip this idea for now. Can be easily added later, if ever needed.

mareklibra replied:
An idea for the future: can this be a set of inputs for a next workflow?

mareklibra commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,48 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Workflow Options",
```
Let's rename to `Workflow Result`

batzionb commented on shared/schemas/workflow-status-schema.json:
```diff
@@ -0,0 +1,83 @@
+                        "type": "string"
+                    },
+                    "link": {
+                        "description": "An URL.",
```
```suggestion
                        "description": "A URL.",
```

mareklibra commented:
Compiling the comments so far:

- Let's not duplicate the status here, keep using the `workflow run's` status (ACTIVE, ERROR, COMPLETED, ABORTED, SUSPENDED, PENDING).
  This result will be filled optionally - when the workflow reports it.
  So renaming the property here to `CompletedWith` to be one of 2 values only (error and success). Considering context. there should be neither pending nor running value.

- Since we removed the old `options`, we can flatten the structure

mareklibra commented on shared/schemas/workflow-ref-schema.json:
```diff
@@ -0,0 +1,20 @@
+    "description": "Workflow reference",
+    "type": "object",
+    "properties": {
+        "id": {
```
Do we need a link here? I guess it can be determined from the ID.

masayag replied:
The ID should be enough. 
Perhaps in the future we'll have to support more than a single Sonataflow stack and then we'll need another identifier for it, but this schema doesn't prevent from extending it.

mareklibra commented on shared/schemas/workflow-option-schema.json:
```diff
@@ -0,0 +1,28 @@
+            "description": "Free form value of the option.",
+            "type": "string"
+        },
+        "valueType": {
```
Or we can provide a property per each type.
Instead of a single `value`, we can have `number`, `text`, `link`. I prefer single `value` and having sort of the `valueType` to guide next processing (e.g. UI decisions)

masayag replied:
This suggestion looks good to me [key,value,valueType].
Should there be a default value for the `valueType` (e.g. string) ?
@batzionb?

ydayagi replied:
I am in favor of having well-defined types and not just arbitrary key-value pairs. if you do go with that option then remove the valueType. Always a string and it is up to the code that uses the key-value to interpret the string.

masayag replied:
The UI will not know to to treat a `string` value that represents a number or URL.

ydayagi replied:
why not? sure it will. based on the name. E.g. key=='my-url', value=='http://cnn.com', key=='my-age', value=='49'. and if we say this is not good then we need to define an exact object

mareklibra replied:
Per offline discussion, we are introducing new optional `format` property to avoid such heuristics. It will add us more options for extension as well.

batzionb commented:
> Compiling the comments so far:
> 
> * Let's not duplicate the status here, keep using the `workflow run's` status (ACTIVE, ERROR, COMPLETED, ABORTED, SUSPENDED, PENDING).
>   This result will be filled optionally - when the workflow reports it.
>   So renaming the property here to `CompletedWith` to be one of 2 values only (error and success). Considering context. there should be neither pending nor running value.
> * Since we removed the old various `*options`, we can flatten the structure

I'm concerned that the completedWith will not always be in sync with the instance status. I think there can be cases where the instance moves to error by sonata. Maybe the UI needs to look at both.

mareklibra commented:
@masayag These "results" are meant to be a part of the `/workflows/instances/:id` response.
Recently, we use `instance.variables` property of the response to hold "the results".

With the new results-structure, will we keep the same location? 
Wdyt about renaming the `instance.variables` to `instance.result` (mind the singular as well).

For reference:
```
{
    "instance": {
        "id": "a332c786-ee29-4524-9497-c87e09b3e81f",
        "processName": "Hello World Workflow",
....REMOVED
        "state": "COMPLETED",
        "start": "2024-09-10T06:25:01.462Z",
        "end": "2024-09-10T06:25:01.46Z",
        "nodes": [
...REMOVED
        ],
        "variables": {
            "workflowdata": {
                "myKey": "myValue",
                "mantra": "Serverless Workflow is awesome!",
                "greeting": "Hello World"
            }
        },
        "error": null,
...REMOVED
        "workflowdata": {
            "myKey": "myValue",
            "mantra": "Serverless Workflow is awesome!",
            "greeting": "Hello World"
        },
        "status": "Completed"
    }
}
```

gciavarrini commented:
@mareklibra 
> @masayag These "results" are meant to be a part of the /workflows/instances/:id response.
Recently, we use instance.variables property of the response to hold "the results".

Here the definition of that [endpoint](https://github.com/janus-idp/backstage-plugins/blob/d1e48db92c3976a471f33133465f1f21d1fddcac/plugins/orchestrator-common/src/openapi/openapi.yaml#L137-L161) (I've already a task to improve the descriptions... sorry about that). Now the `variables` var doesn't exist anymore, it's [workflowdata](https://github.com/janus-idp/backstage-plugins/blob/d1e48db92c3976a471f33133465f1f21d1fddcac/plugins/orchestrator-common/src/openapi/openapi.yaml#L477).
Open to change the spec for this If needed.

masayag commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+                "$ref": "workflow-ref-schema.json"
+            }
+        },
+        "options": {
```
I'm not convinced the `options` is the best name to describe this property.
Perhaps `outputs` or `resultDetails` or a different name that better fits this property?

mareklibra replied:
The purpose is to hold data produced by the workflow. So the `outputs` sounds best here.

masayag commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+            "description": "Additional structured output of workflow processing. This can contain identifiers of created resources, links to resources, logs or other output.",
+            "type": "array",
+            "items": {
+                "$ref": "workflow-option-schema.json"
```
let's rename this file name according to the name that will replace the `options` property above.

mareklibra commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+    "description": "Results of a workflow execution",
+    "type": "object",
+    "properties": {
+        "completedWith": {
```
As @batzionb mentioned, we probably don't need this property at all. The workflow developer should set the workflow status to either success or error and so this is redundant.

I would keep the optional `description` property below.

mareklibra replied:
After a discussion: keeping/removing the `completedWith` depends on the ability to use `error` field in the instance. Let's keep that decision open/

ydayagi commented on shared/schemas/workflow-option-schema.json:
```diff
@@ -0,0 +1,28 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "WorkflowOption",
```
I think the name needs to be changed. what does workflow option mean?

mareklibra replied:
done. We use `outputs` now

masayag commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+        }
+    },
+    "required": [
+        "completedWith"
```
My concern about this if this schema is going to be used while a workflow is still running or only at the end of it.
Since while the workflow is running, it may produce result items that we can/should render in the UI.
However, since the workflow is still running, there is no value for `completedWith` at that time.
@batzionb how do we fetch the workflow results ATM, while the workflow is still running?

mareklibra replied:
Per comment above, the `completedWith` will be removed.

batzionb replied:
> My concern about this if this schema is going to be used while a workflow is still running or only at the end of it.
> Since while the workflow is running, it may produce result items that we can/should render in the UI.
> However, since the workflow is still running, there is no value for completedWith at that time.
> @batzionb how do we fetch the workflow results ATM, while the workflow is still running?

This is a good point. I see it more as a UX question. While workflow is running, the results card should have some kind of busy state or not be shown at all. Not sure how the API should be effected by this if at all.

ydayagi commented on shared/schemas/workflow-option-schema.json:
```diff
@@ -0,0 +1,28 @@
+        },
+        "valueType": {
+            "description": "More detailed type of the 'value' property. Defaults to 'text'.",
+            "enum": [
```
I think it is wrong to have a free form type with just key-value pairs but try to enforce a type on each value.

ydayagi commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+                "$ref": "workflow-ref-schema.json"
+            }
+        },
+        "options": {
```
I don't understand what is this field/object for. if it is for an unknown piece of data then I suggest removing it. the results should be a well-defined object.

gabriel-farache replied:
M2K has no real output per se: the questionnaire link is useful during the execution but in the end, the output of M2K is what is in the target git repo and branch which is provided as input

ydayagi replied:
the exising examples are all links. so we do see there's a unified format. we can limit the output we expect. we can say that the output will have messages and links and that's it

mareklibra replied:
Per offline discussion, a list with priorities is what we introduce.
In the future, we might add a field to clasify the suggested workflows (like recommended, optional, etc)

ydayagi commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+            "description": "High-level summary of the current status, free-form text, human readable.",
+            "type": "string"
+        },
+        "nextWorkflows": {
```
do we have a use case for more than one workflow? i suggest limiting the implementation to the current needs.

ydayagi commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Workflow Results",
```
the results object i created some time ago had a boolean for success/fail and a message. I think this is what we need. in my opinion the enum only complicates things.

mareklibra replied:
After a follow-up discussion, let's investigate how we can reuse instance's `error` field. If there is no simple way, we will need the `completedWith`. Otherwise, let's remove the `compledWith` and use just the `message` property.

masayag commented:
> With the new results-structure, will we keep the same location?
Wdyt about renaming the instance.variables to instance.result (mind the singular as well).

IIUC, this is the result object returned from Sonataflow without any processing. 
It will be the backend's responsibility to return a proper response.

masayag commented on shared/schemas/workflow-options-output-schema.json:
```diff
@@ -0,0 +1,37 @@
+{
```
@mareklibra pls rename this file as well.
Seems this file should be named `workflow-outputs-schema.json` 
and the second file should be named in a singular form `workflow-output-schema.json`

mareklibra replied:
Done. What about `workflow-output-schema.json` and `workflow-output-schema.json`?

masayag replied:
sounds good. i was thinking of both, but then I noticed it isn't clear if the result contains the output or the opposite way around.
perhaps `workflow-result-schema.json` and `workflow-result-item-schema.json` so it is clear which schema is the main one.

masayag commented on shared/schemas/workflow-result-schema.json:
```diff
@@ -0,0 +1,37 @@
+        }
+    },
+    "required": [
+        "completedWith"
```
I'm not sure about having this as required. My concern is about being too enforcing.

mareklibra replied:
ok, removing
